### ISSUE TITLE
content picker not loaded on ubuntu touch

### ISSUE THREAD

**Issue by LiohMoeller** [kiwix/kiwix-js#1279]
On Ubuntu Touch it is not possible to add a zim file, as the content picker does not show up on a click on the 'Load ZIM' or 'Select Folder' Button.

**Comment by Jaifroid** [comment]
@LiohMoeller @kelson42 The problem is that I don't have an Ubuntu Touch device to test on. I've no idea why the API has broken. I wonder if it's been restricted recently? We used to have someone willing to test, and we'd send them a package before release, they'd test it and also diagnose any issues. For the last several releases, we've been releasing blind for this platform. I'm not sure what I could do about it unless someone with a device were willing to take a look at the code and debug, or at least point me in the right direction. Possibly the File System Access API appears to be supported but isn't actually, and the fallback to the old File Picker isn't working.

**Comment by LiohMoeller** [comment]
I can just send/donate you a device for development, if that would help.

**Comment by Jaifroid** [comment]
@LiohMoeller That's very kind, but let me see if there's a reliable simulator I could use first, as I think testing on real hardware would be a bit of a last resort (testing time is already quite onerous). I'll come back to you once I've done some research on this.

**Comment by Jaifroid** [comment]
So it seems there is this Ubuntu Touch development kit: https://github.com/ubports/ubports-pdk . I'll see if I can get that to run in order to reproduce the issue.

**Comment by nahcol10** [comment]
I am a Ubuntu user . I can help you to test it.

**Comment by AyushSahoo19** [comment]
Hi team,

Before proceeding, I wanted to ask if the team has already tested the content picker issue using the [Ubports PDK](https://github.com/ubports/ubports-pdk) to simulate the environment?

Since I don’t have access to an Ubuntu Touch device, my plan is to start by setting up this PDK to simulate the environment. This should help me:

1. Replicate the Issue: I’ll focus on simulating the API behavior and how the content picker interacts with the system.
2. Test Fallbacks: I’ll check whether the fallback to the old File Picker is functioning properly, especially if the File System Access API is unsupported or restricted.
If I’m unable to reproduce the issue or need further testing on a real device, would it be possible to collaborate with someone who has access to an actual Ubuntu Touch device? Alternatively, do you have any recommendations for other reliable emulators or testing environments beyond the PDK?

Once I have the PDK set up, I’ll begin debugging the content picker and share my findings. Let me know if you have any specific areas in the codebase you think I should focus on, or any additional guidance for testing.

**Comment by Jaifroid** [comment]
Also, the fact that there is a "Select folder" button strongly suggests my hypothesis is true. One of these APIs is being detected, but something is breaking. If neither of those APIs were detected, then the system would show the HTMl5 file picker.

**Comment by AyushSahoo19** [comment]
Thanks for the details! Could you point me to the specific file or section of the codebase where the content picker’s behavior and fallback mechanisms are handled? This will help me focus on the right part while testing the bug on the Ubuntu Touch emulator. Let me know, and I’ll dive in!

**Comment by Jaifroid** [comment]
Yes of course, just search for `openLocalFiles` in your code editor, and you should find the picker in `index.html` and `app.js` (where the initial logic is handled). There is also `abstractFilesystemAccess.js` which deals with the logic for the File System Access API, `webkitdirectory` and a few other things.

**Comment by AyushSahoo19** [comment]
Before I start debugging the content picker issue, could you let me know which version of Ubuntu (or Ubuntu Touch) I should focus on testing this for? This will help ensure I'm replicating the correct environment.

**Comment by Jaifroid** [comment]
I updated the Ubuntu Touch SDK to 20.04 for the last release. I have not had any feedback as to whether the file picker is working or not under that SDK, despite there appearing to be a number of Ubuntu touch users of this app. It's long enough to assume that people would complain if the issue were still ongoing. Feel free to reopen if not fixed.

**Comment by garagecoder** [comment]
@Jaifroid Hi, sadly this is still a problem (and in fact always has been on U.T.) at least for me. Using Touch 24.04 now and the "Select ZIM file(s)" button registers a press but does nothing. I'll try to contribute more than just this message, if I get the time!

**Comment by ChromiumOS-Guy** [comment]
yeah, please hardcode it to the html5 picker in UT for now, a solution is better then no solution.

as for support once I'm done with my exam season I'll continue working on XDG desktop portals support if that happens XDG apps will be possible next step will be flatpaks from there its quite possible that you will be able to drop support completely but thats a long way's off for now.

**Comment by Jaifroid** [comment]
OK, I'm sorry to hear this is still an issue (and I missed that you said so in November), even after updating to new SDK. The picker detects the best supported API (the API that the system says is supported) and uses that. Most likely it's trying to use the File System Access API, but in fact it's not supported even though the relevant methods are available. The problem is that without access to an Ubuntu Touch device or an accurate simulator, I can't test what actually works on UT. If you have a device, especially if you can debug from a connected PC, then it would help immensely to see what the cause of failure is.

**Comment by ChromiumOS-Guy** [comment]
you don't need an accurate method only lomiri which installable on a debian trixie with a simple ```sudo apt install lomiri```
you should also install all base apps, I don't know if theres a package that does that but everything important can be installed with a ```sudo apt install lomiri-*``` though its a bit excessive.

from there you can install kiwix through openstore and you'll have the same exact problem, also run ```clickable desktop``` if you haven't already if it has the bug there then it will be much easier to debug.

as for why you only need lomiri, if you install only lomiri on a headless install of debian, it won't have whatever picker APIs that you support, it will only have what lomiri has which is the DE for ubuntu touch.

**Comment by ChromiumOS-Guy** [comment]
as for me testing specifically I daily drive a UT device so that won't be a problem, I'm close to the end of exam season, i'll have my last test for good month or so at the 17th this month, and while my main focus in implementing XDG desktop portals into Ubuntu Touch I will take a look then but it will take me a while and you should not count on it because there is still a lot of time management involved just to clear time for the XDG stuff.

**Comment by Jaifroid** [comment]
Thanks, that's very helpful, I'll check it out. It will be very useful to have an actual UT user in the loop, and of course one who knows the innards so well.

**Comment by ChromiumOS-Guy** [comment]
ubports-pdk has been fixed, if it still does not work you can setup a script with this: gitlab.com/-/snippets/5937745
new documentation on how to get a virtual machine with lomiri is being created as discussed in recent devsync, as all documentation is outdated or missing.

**Comment by Jaifroid** [comment]
@ChromiumOS-Guy Thank you very much for the info. I'll get onto it asap, as I definitely want to get this blocker fixed.
